prboom2/src/MUSIC/portmidiplayer.c

```diff
@@ -463,36 +404,44 @@ static dboolean is_sysex_reset (unsigned char *data)
 
 static void writesysex (unsigned long when, int etype, unsigned char *data, int len)
 {
-  // sysex code is untested
-  // it's possible to use an auto-resizing buffer here, but a malformed
-  // midi file could make it grow arbitrarily large (since it must grow
-  // until it hits an 0xf7 terminator)
+  // sysex messages in midi files (smf 1.0 pages 6-7):
+  // complete:        (F0 ... F7)
+  // multi-packet:    (F0 ...) + (F7 ...) + ... + (F7 ... F7)
+  // escape sequence: (F7 ...)
+
   if (len + sysexbufflen > SYSEX_BUFF_SIZE - 1)
   {
-    lprintf (LO_WARN, "portmidiplayer: ignoring large or malformed sysex message\n");
+    // ignore messages that are too long
     sysexbufflen = 0;
     return;
   }
+
+  if (etype == MIDI_EVENT_SYSEX_SPLIT && sysexbufflen == 0)
+  {
+    // ignore escape sequence
+    return;
+  }
+
+  if (etype == MIDI_EVENT_SYSEX)
+  {
+    // start a new message (discards any previous incomplete message)
+    sysexbuff[0] = MIDI_EVENT_SYSEX;
+    sysexbufflen = 1;
+  }
+
   memcpy (sysexbuff + sysexbufflen, data, len);
   sysexbufflen += len;
-  if (sysexbuff[sysexbufflen - 1] == 0xf7) // terminator
-  {
-    memmove(&sysexbuff[1], &sysexbuff[0], sysexbufflen * sizeof(*sysexbuff));
-    sysexbuff[0] = 0xf0; // start of exclusive (SOX) in front
-    sysexbufflen++;
 
-    #ifdef _WIN32
-    if (!mus_extend_volume)
-    #endif
+  // process message if it's complete, otherwise do nothing yet
+  if (sysexbuff[sysexbufflen - 1] == MIDI_EVENT_SYSEX_SPLIT)
+  {
+    if (is_mastervol(sysexbuff, sysexbufflen))
```

Comment:
Are there files with master volume events? Maybe we should just ignore these events?

Reply:
Yes there are. If for some reason a midi file wants to adjust the volume via master volume, it doesn't hurt to scale it accordingly. I know of two more vendor specific master volume messages (mentioned in Roland and Yamaha manuals) but I haven't seen them in midi files. Actually, I could probably add a check for those just to be safe.